src/v/ssx/thread_worker.h

```diff
@@ -44,13 +50,36 @@ class task_base {
     task_base& operator=(task_base&&) = delete;
     task_base& operator=(task_base const&) = delete;
 
-    virtual void process(ss::alien::instance&, ss::shard_id) = 0;
+    virtual ss::stop_iteration process(ss::alien::instance&, ss::shard_id) = 0;
     virtual void
     set_exception(ss::alien::instance&, ss::shard_id, std::exception_ptr)
       = 0;
     virtual ~task_base() = default;
 };
 
+class stop_task final : public task_base {
+public:
+    ss::future<> get_future() noexcept { return _promise.get_future(); }
+
+    ss::stop_iteration
+    process(ss::alien::instance& alien, ss::shard_id shard) final {
+        ss::alien::run_on(
+          alien, shard, [this]() noexcept { _promise.set_value(); });
```

Comment:
presumably the promise here is being accessed by alien thread and reactor thread? if it is, that doesn't strike me as safe. is this a normal pattern working with alien thread?

Reply:
No, this is running on the reactor. It's confusingly named, but `ss::alien::run_on` runs some code *on the reactor* from an alien thread. This is the same thing we do below in the "normal" task fwiw